gym/spaces/graph.py

```diff
@@ -0,0 +1,260 @@
+"""Implementation of a space that represents graph information where nodes and edges can be represented with euclidean space."""
+from typing import Optional, Sequence, Union
+
+import numpy as np
+
+from gym.spaces.box import Box
+from gym.spaces.discrete import Discrete
+from gym.spaces.multi_discrete import MultiDiscrete
+from gym.spaces.space import Space
+from gym.utils import seeding
+
+
+class GraphObj:
+    r"""A base for constructing information as graphs."""
+
+    def __init__(
```

Comment:
I haven't done much with graphs in a long time, but can they not be represented by just an adjacency matrix and a node list (I guess, though not necessary)? Have I missed something, but how can be a graph be represented by just a list of nodes?

Reply:
Yes they can, but that comes with 2 immediate downsides - 1. n^2 instead of 2n memory complexity (where n can be m^2 for m nodes), 2. Inability to represent directional edges without having two adjacency matrices (or one with different tokens representing different connections, which further complicates things). Though it is a much rarer use case, an adjacency matrix also can't represent multiple edges between two nodes.

Using a list makes things much easier for computation using GNNs or transformer-like models as well, since most of the time the `edge_links` can just be concatenated on top of the `edges` and directly passed to the model.

In any case, one should be able to construct an adjacency matrix using `edge_links` fairly easily since it's already a list of indices.

Reply:
In any case, if an adjacency matrix is the preferred way to do things, I propose doing so using a wrapper instead.